---
title: Tutorial of InstaPrism
author: Mengying Hu
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    theme: lumen
    highlight: default
    toc: true
    toc_depth: 2
toc-title: "Table of Contents"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction
**InstaPrism** is R package to deconvolute cellular proportion and gene expression in bulk RNA-Seq data. Based on the same conceptual framework and corresponding generative mode from [BayesPrism](https://github.com/Danko-Lab/BayesPrism), **InstaPrism** re-implements BayesPrism in a derandomized framework by replacing the time-consuming Gibbs sampling steps in BayesPrism with a fixed-point algorithm, which greatly accelerated the calculation speed while maintaining highly comparable performance. \

In this tutorial, we provide three examples of running **InstaPrism** and compare the results with BayesPrism.

## Getting started

Load the **InstaPrism** package. \

**InstaPrism** works as an independent R package and does not require the users to have BayesPrism installed. \
We recommend the readers to read through the [tutorial](https://github.com/Danko-Lab/BayesPrism/blob/main/tutorial_deconvolution.pdf) of BayePrism before running our examples. \
`extdata` folder associated with this tutorial can be downloaded from [zenodo](https://zenodo.org/record/7730658) repository. 

```{r setup}
library(InstaPrism)
```

## Example 1: deconvolution on a small simulated dataset
\

#### Step 1. Create simulated single cell and bulk expression data
The following lines simulate bulk and single-cell expression, as well as marker genes and true proportions that can be used as an example of deconvolution using the simulation model the **BisqueRNA** package. 

```{r,eval=FALSE}
library(Biobase)
library(BisqueRNA)
cell.types <- c("Neurons", "Astrocytes", "Oligodendrocytes", "Microglia", "Endothelial Cells","others")
avg.props <- c(.5, .2, .15, .07, .03,.05)
sim.data <- SimulateData(n.ind=20, n.genes=1000, n.cells=500, cell.types=cell.types, avg.props=avg.props)
```

sim.data contains the following simulated objects:

* A 1000 $\times$ 10000 single-cell expression object for 20 individuals, with cells annotated with inidivual ID and cell type labels.
* A 1000 $\times$ 20 bulk expression object for 20 individuals.
* A 6 $\times$ 20 matrix indicating cellular proportions for 20 individuals.
* A dataframe indicating marker genes for each cell type

Alternatively, readers can load the example simulated data directly from the **InstaPrism** package.
```{r}
data("sim.data")
```
\

#### Step 2. Prepare input for InstaPrism and BayesPrism
Both **InstaPrism** and BayesPrism take single cell profiles as prior information and return deconvolution results for a given bulk expression matrix. To prepare input for InstaPrism and BayesPrism, we need:

* scExpr: a single-cell expression data as prior information (raw count or CPM normalized expression)
* bulk_Expr: a bulk expression to run deconvolution (raw count or TPM normalized expression)
* cell.type.labels: a character vector indicating cell types of each cell from the scRNA data
* cell.state.labels: a character vector indicating cell states of each cell from the scRNA data

In real practice, cell.state.labels usually denote different cell states from a same given cell type. For example, malignant cells can be subclustered by different patients to denote different malignant states. With the *sim.data* we created above, we will generate some artificial cell state labels for *Neurons* given individual IDs.


```{r}
library(Biobase)
sc.eset <- sim.data$sc.eset
bulk.eset <- sim.data$bulk.eset

sc_Expr = exprs(sc.eset)
sc_Expr = apply(sc_Expr,2,function(x)((x/sum(x))*1e+05)) 

bulk_Expr = exprs(bulk.eset)
bulk_Expr = apply(bulk_Expr,2,function(x)((x/sum(x))*1e+06)) 

cell_type_labels=sc.eset@phenoData@data[["cellType"]]

# create artifical cell-state labels for Neuron cells: Neuron_A, Neuron_B, Neuron_C
Neurons_states=names(table(sc.eset@phenoData@data$SubjectName))
names(Neurons_states)=c(rep("A",8),rep("B",6),rep("C",6))

cell_state_labels=ifelse(sc.eset@phenoData@data[["cellType"]]=='Neurons',paste0('Neurons_sub',names(Neurons_states)[match(sc.eset@phenoData@data$SubjectName,Neurons_states)]),sc.eset@phenoData@data[["cellType"]])

table(cell_state_labels)
```
\

#### Step 3. Run deconvolution with InstaPrism and BayesPrism 
\ 
`InstaPrism()` is the main deconvolution function. Depending the format of single cell prior information to use, users need to specify the `input_type` argument passed to the function. The `InstaPrism()` function currently support **four** input_types.  \
Note that in the following examples we will consider a simple implementation where we use prior information from the scRNA Seq directly, without the need to update reference.

**1. InstaPrism with `input_type = 'raw'`**

Under this setting, InstaPrism takes the same input format as BayesPrism:

* scExpr: a single-cell expression data as prior information (raw count or CPM normalized expression)
* bulk_Expr: a bulk expression to run deconvolution (raw count or TPM normalized expression)
* cell.type.labels: a character vector indicating cell types of each cell from the scRNA data
* cell.state.labels: a character vector indicating cell states of each cell from the scRNA data

```{r,results='hide'}
library(tictoc) # load package to measure execution time
tic()
InstaPrism.res = InstaPrism(input_type = 'raw',sc_Expr = sc_Expr,bulk_Expr = bulk_Expr,
                    cell.type.labels = cell_type_labels,cell.state.labels = cell_state_labels,
                    return.Z.cs = T, return.Z.ct  = T)
```
```{r}
elapsed1 = toc()
```

**2. InstaPrism with `input_type = 'prism'`**

In this mode, users need to first construct a prism object using the BayesPrism package, and then pass the `prismObj` to the `InstaPrism()` function. 

```{r,results='hide'}
library(BayesPrism)
tic()
bp.obj = new.prism(reference = t(sc_Expr),input.type = 'count.matrix',
                   cell.type.labels = cell_type_labels, cell.state.labels = cell_state_labels,
                   mixture = t(bulk_Expr),key = NULL)
InstaPrism.res2 = InstaPrism(input_type = 'prism',prismObj = bp.obj,return.Z.cs = T,return.Z.ct  = T)
```
```{r}
elapsed2 = toc()
```

**3. InstaPrism with `input_type = 'refPhi'` (recommended)**

Under this setting, users need to pass a pre-defined `refPhi` object containing single cell prior information. To build a `refPhi` object, users can use the `refPrepare()` function with the following arguments:

* scExpr: a single-cell expression data as prior information (raw count or CPM normalized expression)
* cell.type.labels: a character vector indicating cell types of each cell from the scRNA data
* cell.state.labels: a character vector indicating cell states of each cell from the scRNA data

Alternatively, users can use the the pre-built `refPhi` object we provided for different tumor types (will come out soon). \
The advantage of using `refPhi` as input is that one doesn't need to load the large single cell expression matrix every time, and the `refPhi` object can be easily re-implemented once built. 

```{r,results='hide'}
refPhi_obj = refPrepare(sc_Expr, cell_type_labels, cell_state_labels, output_class = 'refPhi')
tic()
InstaPrism.res3 = InstaPrism(input_type = 'refPhi',bulk_Expr = bulk_Expr,refPhi = refPhi_obj, return.Z.cs = T,return.Z.ct  = T)
```
```{r}
elapsed3 = toc()
```

**4. InstaPrism with `input_type = 'refPhi_cs'` (advanced practice)**

Using `refPhi_cs` provides a more versatile setting where users can selectively choose which cell-states to use for deconvolution. \
Given a pre-built `refPhi` object, users can modify it easily by selectively include cell-states for evaluation. For example, given the `refPhi` object we built in the previous section, if we only want to include `Neurons_subA` and `Neurons_subB` as cell states of interest, and remove `Neurons_subC` from Neurons, we could: 

```{r,results='hide'}
refPhi_cs_map = refPhi_obj@map
refPhi_cs_map$Neurons = c('Neurons_subA', 'Neurons_subB' )
to_move_id = which(colnames(refPhi_obj@phi.cs) == 'Neurons_subC')
refPhi_cs_obj = new('refPhi_cs',phi.cs = refPhi_obj@phi.cs[,-to_move_id],map = refPhi_cs_map)

InstaPrism.res4 = InstaPrism(input_type = 'refPhi_cs',bulk_Expr = bulk_Expr,refPhi_cs = refPhi_cs_obj, return.Z.cs = T,return.Z.ct  = T)
```
However, note that the `updateReference` module will not be available under this setting, since we are missing cell.type specific reference phi, which is required to update reference. \
\

**Run BayesPrism for comparison**

In the following example, we are setting `update.gibbs = F` for consistency with our previous examples where we only consider single-cell based reference, without updating the reference. Deconvolution with BayesPrism can be called by a single function `run.prism()` as follows:

```{r,eval=FALSE}
start.time = Sys.time()
bp.res = run.prism(bp.obj,update.gibbs = F)
end.time=Sys.time()
bp_running_time = difftime(end.time, start.time,units = 'secs') %>% as.numeric()
save(bp.res,bp_running_time,file='extdata/tutorial_example1/bp.res.initial.RData')
```
\ 

```{r,include=FALSE}
load('extdata/tutorial_example1/bp.res.initial.RData')
```

#### Step 4. Evaluate deconvolution results of InstaPrism and BayesPrism 

* deconvolution performance of InstaPrism

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We provide a useful function `deconv_performance_plot()` that visualize correlations at per-cell.type level. From the plot, InstaPrism accurately predicts cellular proportions from sim.data

```{r,fig.align='center',fig.width=10,fig.height=2.35}
deconv_performance_plot(est = InstaPrism.res@Post.ini.ct@theta,true = sim.data$props,title = 'InstaPrism performance on sim.data',nrow=1)
```

* compare cell type fraction estimates from InstaPrism and BayesPrism

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We can visualize the correlation between two fraction estimates with a heatmap. As shown by the heatmap, cell.type estimates from both methods are highly correlated 

```{r}
corr=cor(t(InstaPrism.res@Post.ini.ct@theta),bp.res@posterior.initial.cellType@theta)
diag(corr)
```
```{r,fig.align='center',fig.height=3.65,fig.width=4.5}
ComplexHeatmap::Heatmap(corr,show_row_dend = F,show_column_dend = F,column_title = 'BayesPrism',row_title = 'InstaPrism',name='correlation')
```


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; note that we did not include a cell.state comparison here because with artificial cell.state.labels imposed on cells that are not intrinsically distinguishable, it will generate bias on the cell.state estimates

* cell.type specific gene expression comparison

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; In the following plots, we showed the correlation between cell,type specific gene expression values (Z matrix) between BayesPrism and InstaPrim for one bulk sample. The plots suggest that the Z matrix estimates by both methods is highly correlated.

```{r,fig.align='center',fig.width=12,fig.height=2.8}
deconv_performance_plot(t(InstaPrism.res@Post.ini.ct@Z[1,,]),t(bp.res@posterior.initial.cellType@Z[1,,]),nrow = 1,title = 'correlation between cell.type specific gene expression \n',xlabel = 'BayesPrism',ylabel = 'InstaPrism')
```


* running time comparison

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; InstaPrism significantly accelerates the deconvolution speed in either modes.

```{r,fig.align='center',fig.width=4.8,fig.height=3.3}
InstaPrism_mode1_running_time = as.numeric(elapsed1$toc-elapsed1$tic)
InstaPrism_mode2_running_time = as.numeric(elapsed2$toc-elapsed2$tic)
InstaPrism_mode3_running_time = as.numeric(elapsed3$toc-elapsed3$tic)

rt=data.frame(method=c('InstaPrism.raw.mode','InstaPrism.prism.mode','InstaPrism.refPhi.mode','BayesPrism'),
              time=c(InstaPrism_mode1_running_time,InstaPrism_mode2_running_time,
                     InstaPrism_mode3_running_time,bp_running_time))
rt$time=round(rt$time,2)
ggplot(rt,aes(method,time))+
  geom_bar(stat="identity",fill='grey',width = 0.5)+
  theme_bw()+
  xlab('')+
  ylab('time (secs)')+
  geom_text(aes(label = time), vjust = -0.2)+
  ylim(0,max(rt$time)*1.05)+
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```

\

#### Step 5 (optional). Run deconvolution with InstaPrism and BayesPrism using the updated reference
We now consider a more advanced deconvolution problem where we want to leverage from the information shared by bulk samples and update the reference matrix accordingly. \

This step is implemented in BayesPrism by setting `update.gibbs = TRUE` in `run.prism()` function, or by calling the `update.theta()` function on the initial deconvoluton object. In **InstaPrism**, we incorporated the `update.theta()` module from BayesPrism and replaced the subsequent Gibbs Sampling step with our fixed-point alogirthm. To enable reference update, simply set `update = T` \

**i. deconvolution with InstaPrism**

```{r,results='hide',warning=FALSE}
tic()
InstaPrism.res.updated = InstaPrism(input_type = 'raw',sc_Expr = sc_Expr,bulk_Expr = bulk_Expr,
                    cell.type.labels = cell_type_labels, cell.state.labels = cell_state_labels,
                    update=T)
```
```{r}
elp = toc()
```
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;With update = T, InstaPrism function returns an S4 object containing the following information:

* Post.ini.cs: posterior information for cell.states
    + Z: a three dimension array indicating gene expression of each cell.state in different individual (returned when `return.Z.cs=T`, default = F)
    + theta: cell.state fraction estimates for each individual
* Post.ini.ct: posterior information for cell.types
    + Z: a three dimension array indicating gene expression of each cell.types in different individual (returned when `return.Z.ct=T`, default = F)
    + theta: cell.types fraction estimates for each individual
* Post.updated.ct: cell.type fraction estimates using updated reference
    + theta: cell.types fraction estimates for each individual with updated refernce
* initial.reference: an S4 object with scRNA based prior information as reference
* updated.reference: an S4 object with updated reference


**ii. deconvolution with BayesPrism**
```{r,eval=FALSE}
start.time = Sys.time()
bp.res.updated = run.prism(bp.obj,update.gibbs = T)
# alternatively, run initial theta estimation and updated theta estimation separately
# bp.res = run.prism(bp.obj,update.gibbs = F)
# bp.res.updated = update.theta(bp=bp.res)
end.time=Sys.time()
bp_updated_running_time = difftime(end.time, start.time,units = 'secs') %>% as.numeric()
save(bp.res.updated,bp_updated_running_time,file = 'extdata/tutorial_example1/bp.res.updated.RData')
```

**iii. deconvolution results comparison**

cell.type estimation from both methods are highly correlated 

```{r,include=FALSE}
load('extdata/tutorial_example1/bp.res.updated.RData')
```

```{r}
corr=cor(t(InstaPrism.res.updated@Post.updated.ct@theta),bp.res.updated@posterior.theta_f@theta)
diag(corr)
```
```{r,fig.align='center',fig.height=3.65,fig.width=4.5}
ComplexHeatmap::Heatmap(corr,show_row_dend = F,show_column_dend = F,column_title = 'BayesPrism',row_title = 'InstaPrism',name='correlation')
```

* running time comparison

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; InstaPrism significantly accelerates the deconvolution speed using the updated reference.

```{r,fig.align='center',fig.width=4.3,fig.height=2.45}
InstaPrism_updated_running_time = as.numeric(elp$toc - elp$tic)
rt=data.frame(method=c('InstaPrism','BayesPrism'),
              time=c(InstaPrism_updated_running_time,bp_updated_running_time))
rt$time=round(rt$time,2)
ggplot(rt,aes(method,time))+
  geom_bar(stat="identity",fill='grey',width = 0.4)+
  theme_bw()+
  ylab('time (secs)')+
  geom_text(aes(label = time), vjust = -0.2)+
  ylim(0,max(rt$time)*1.05)
```

## Example 2: deconvolution on the tutorial data from BayesPrism
\
In this example, we will use the tutorial [data](https://github.com/Danko-Lab/BayesPrism/tree/main/tutorial.dat) provided in BayesPrism and compare the deconvolution results between InstaPrism and BayesPrism. \

#### Step 1. Run BayesPrism following the [tutorial](https://github.com/Danko-Lab/BayesPrism/blob/main/tutorial_deconvolution.pdf) in BayesPrism. \

Note that it takes more than 6 hours to run the following code from our end (using n.core=16), users can skip the BayesPrism running process by loading our processed results directly. 

```{r,eval=FALSE}
library(BayesPrism)
load('extdata/tutorial_example2/tutorial.gbm.rdata')
sc.stat <- plot.scRNA.outlier(
  input=sc.dat, #make sure the colnames are gene symbol or ENSMEBL ID 
  cell.type.labels=cell.type.labels,
  species="hs", #currently only human(hs) and mouse(mm) annotations are supported
  return.raw=TRUE #return the data used for plotting. 
  #pdf.prefix="gbm.sc.stat" specify pdf.prefix if need to output to pdf
)
bk.stat <- plot.bulk.outlier(
  bulk.input=bk.dat,#make sure the colnames are gene symbol or ENSMEBL ID 
    sc.input=sc.dat, #make sure the colnames are gene symbol or ENSMEBL ID 
  cell.type.labels=cell.type.labels,
  species="hs", #currently only human(hs) and mouse(mm) annotations are supported
  return.raw=TRUE
  #pdf.prefix="gbm.bk.stat" specify pdf.prefix if need to output to pdf
)
# Filter outlier genes from scRNA-seq data
sc.dat.filtered <- cleanup.genes (input=sc.dat,
                                  input.type="count.matrix",
                                    species="hs", 
                                    gene.group=c( "Rb","Mrp","other_Rb","chrM","MALAT1","chrX","chrY") ,
                                    exp.cells=5)
# Subset protein coding genes
sc.dat.filtered.pc <-  select.gene.type (sc.dat.filtered,
                                        gene.type = "protein_coding")
# construct a prism object
myPrism <- new.prism(
  reference=sc.dat.filtered.pc, 
  mixture=bk.dat,
  input.type="count.matrix", 
  cell.type.labels = cell.type.labels, 
  cell.state.labels = cell.state.labels,
  key="tumor",
  outlier.cut=0.01,
    outlier.fraction=0.1,
)
# run BayesPrism
start.time = Sys.time()
bp.res <- run.prism(prism = myPrism, n.cores=16,update.gibbs=T) # set update.gibbs=T for full comparison between two methods
end.time=Sys.time()
bp_running_time = difftime(end.time, start.time,units = 'mins') %>% as.numeric()
save(bp.res,bp_running_time,file = 'extdata/tutorial_example2/bp.res.RData')
```
\
Alternatively, users can download the processed results directly from [zenodo](https://zenodo.org/record/7730658) repository
```{r}
load('extdata/tutorial_example2/bp.res.RData')
```
\


#### Step 2. Run InstaPrism 

For full comparison between BayesPrism and InstaPrism, set `update=T` to get cell.type fraction estimates with updated reference. Meanwhile, we will set `n.core = 16` for parallel calculation and `snowfall.ncore = 16` for parallel reference update. \

As an additional note, we recommend the users to set `return.Z.cs = F` when `update=T`, this will prevent potential out-of-memory issues.\

```{r,eval=FALSE}
library(BayesPrism)
start.time = Sys.time()
InstaPrism.res = InstaPrism(input_type = 'prism',prismObj = bp.res@prism,update=T,n.iter = 100,n.core=16,snowfall.ncore = 16) 
# alternatively, users can use myPrism (constructed in the previous example) as a substitute of bp.res@prism; or run InstaPrism under 'raw' mode by specifying corresponding arguments
end.time=Sys.time()
InstaPrism_running_time = difftime(end.time, start.time,units = 'mins') %>% as.numeric()
```

For comparison of initial fraction estimates only, set `update = F`. (Note that by setting `update=F`, we no longer have the `Post.updated.ct` object.)\

```{r,eval=FALSE}
start.time = Sys.time()
InstaPrism.res2 = InstaPrism(input_type = 'prism',prismObj = bp.res@prism, update=F, n.iter = 100, n.core=16) 
end.time=Sys.time()
InstaPrism_running_time2 = difftime(end.time, start.time,units = 'mins') %>% as.numeric()
save(InstaPrism.res,InstaPrism_running_time,InstaPrism_running_time2,file = 'extdata/tutorial_example2/InstaPrism.res.RData')
```

Alternatively, users can download our processed results
```{r,warning=FALSE}
load('extdata/tutorial_example2/InstaPrism.res.RData')
```
\

#### Step 3. Compare deconvolution results

* cell.state fraction comparison 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The tutorial [data](https://github.com/Danko-Lab/BayesPrism/tree/main/tutorial.dat) provided in BayesPrism contains 73 different cell.state.labels. Here we provide an example of how to visualize correlation at per cell.state level, ordered by different categories of cell.states. \

We can find that across 73 different cell.states, most of the fraction estimates are highly correlated. \

```{r,fig.align='center',fig.width=5.7,fig.height=3.9}
# organize different cell.states in the tutorial data
PJ.cs=data.frame(cs=rownames(InstaPrism.res@Post.ini.cs@theta))
PJ.cs$group=ifelse(grepl('tumor',PJ.cs$cs),'malignant','immune/others')
PJ.cs$tumor=ifelse(PJ.cs$group=='malignant',sub("\\-.*", "", PJ.cs$cs),'immune/others')
PJ.cs$tumor=ifelse(PJ.cs$group=='malignant',paste0(PJ.cs$tumor,'-malignant'),'immune/others')
PJ.cs=PJ.cs[order(PJ.cs$group,decreasing = T),]

# get correlation between cell.state fraction estimates
cs.corr=cor(t(InstaPrism.res@Post.ini.cs@theta[PJ.cs$cs,]),bp.res@posterior.initial.cellState@theta[,PJ.cs$cs])

# visualize
library(RColorBrewer)
ha=ComplexHeatmap::HeatmapAnnotation(cell.state.category=PJ.cs$tumor,col =list(cell.state.category=setNames(brewer.pal(9,'Set3'),unique(PJ.cs$tumor))))
ComplexHeatmap::Heatmap(cs.corr,show_row_dend = F,show_column_dend = F,column_title = 'BayesPrism',row_title = 'InstaPrism',name='correlation',cluster_rows = F,cluster_columns = F,show_row_names = F,show_column_names = F,top_annotation = ha)
```
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Note that some deviated cell.state estimations occurs specifically within the same patient ID, suggesting that the assigned cell.states for these cells may not be distinguishable enough, making the subsequent cell.state fraction estimation less separable.\

* cell.type fraction comparison (with initial reference) 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;While there's inconsistency between two methods at per cell.state levels, by aggregating cell.state to cell.type levels, two methods provide exactly the same deconvolution results. 
```{r,fig.align='center',fig.height=3.4,fig.width=4.25}
ct.corr=cor(t(InstaPrism.res@Post.ini.ct@theta),bp.res@posterior.initial.cellType@theta) %>% round(1)

ComplexHeatmap::Heatmap(ct.corr,show_row_dend = F,show_column_dend = F,column_title = 'BayesPrism',row_title = 'InstaPrism',name='correlation', cell_fun = function(j, i, x, y, w, h, col) {grid::grid.text(ct.corr[i, j], x, y)})
```

* running time comparison

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;For this tutorial [data](https://github.com/Danko-Lab/BayesPrism/tree/main/tutorial.dat), InstaPrism shortened the running time from hours to only minutes. The process can be even accelerated by setting `udpate=F` in the `InstaPrism()` function. 

```{r,fig.align='center',fig.width=4.7,fig.height=2.45}
rt=data.frame(method=c('InstaPrism \n (n.core=16)','InstaPrism \n (with update=F,\n n.core=16)','BayesPrism \n (n.core=16)'),
              time=c(InstaPrism_running_time,InstaPrism_running_time2,bp_running_time))
rt$time=round(rt$time,2)
ggplot(rt,aes(method,time))+
  geom_bar(stat="identity",fill='grey',width = 0.5)+
  theme_bw()+
  ylab('time (mins)')+
  geom_text(aes(label = time), vjust = -0.2)+
  ylim(0,max(rt$time)*1.08)
```

* cell.type fraction comparison (with updated reference)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We note that InstaPrism does not produce exactly the same cell.type fraction estimation with the updated reference, this is because the updated reference is sensitive to the Z matrix values, affecting the subsequent cell.type fraction estimation. \
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;We recommend the readers to implement InstaPrism by setting `update=F` in real practice, as it's less time-consuming while maintaining good performance. 
```{r}
ct.corr.updated=cor(t(InstaPrism.res@Post.updated.ct@theta),bp.res@posterior.theta_f@theta)
diag(ct.corr.updated)
```


## Example 3: deconvolution on heterogeneouly simulated bulk data
We have recently proposed a heterogeneous bulk simulation pipeline that simulate bulk samples with realistic biological variance (for more details, check [here](https://github.com/humengying0907/deconvBenchmarking)). In the following example, we will run deconvolution on a heterogeneouly simulated bulk dataset of HNSC tumors and evaluate the deconvolution performance by comparing the fraction estimates with real fractions.

#### Step 1. Load example data

For details about how the bulk samples are simulated, check [here](https://github.com/humengying0907/deconvBenchmarking).

```{r}
load('extdata/tutorial_example3/example3.RData')
```
The `example3.RData` contains the following objects:

* simulated_bulk: simulated bulk expression of HNSC tumors to run deconvolution
* simulated_frac: simulated cell.type fractions for the bulk samples as ground truth
* scExpr_train: sc-RNA data as prior information
* cell.type.labels: a character vector indicating cell types of each cell from the scRNA data
* cell.state.labels: a character vector indicating cell states of each cell from the scRNA data

where we assign the malignant cells with different cell.states by their patient IDs. \


#### Step 2. Run InstaPrism
```{r,results='hide'}
InstaPrism.res = InstaPrism(input_type = 'raw',sc_Expr = scExpr_train,bulk_Expr = simulated_bulk,
                            cell.type.labels = cell.type.labels, cell.state.labels = cell.state.labels, 
                            n.iter = 100,n.core = 16)
```
\

#### Step 3. Compare InstaPrism results with real cell.type fractions
```{r,fig.align='center',fig.width=10,fig.height=4.5,message=FALSE}
deconv_performance_plot(est = InstaPrism.res@Post.ini.ct@theta,true = t(simulated_frac),title = 'InstaPrism performance on heterogenously simulated bulk data',nrow=2)
```
InstaPrism achieves reasonable cell.type estimates for the simulated bulk samples.\


#### Step 4 (optional). Compare with BayesPrism result
We provide an example below showing how to run BayesPrism using the heterogeneously simulated bulk data. 

```{r,eval=FALSE}
myPrism <- new.prism(
  reference=t(scExpr_train), 
  mixture=t(simulated_bulk),
  input.type="count.matrix", 
  cell.type.labels = cell.type.labels, 
  cell.state.labels = cell.state.labels,
  key="malignant",
  outlier.cut=0.01,
  outlier.fraction=0.1,
)
bp.res <- run.prism(prism = myPrism, n.cores=16)
save(bp.res,file = 'extdata/tutorial_example3/bp.res.RData')
```

Alternatively, users can load our processed results directly.
```{r}
load('extdata/tutorial_example3/bp.res.RData')
```

Evaluate BayesPrism performance (with initial reference)
```{r,fig.align='center',fig.width=10,fig.height=4.5,message=FALSE}
deconv_performance_plot(est = t(bp.res@posterior.initial.cellType@theta),true = t(simulated_frac),title = 'BayesPrism performance on heterogenously simulated bulk data \n (using initial reference)',nrow=2)
```
We can find that cell.type fraction estimates from InstaPrism and BayesPrism are highly coordinated. \
\
As an extension, we provide an example of how to compare fraction estimates with initial reference and updated reference.
```{r,fig.align='center',fig.width=5,fig.height=3}
library(tibble)
initial.corr=diag(cor(bp.res@posterior.initial.cellType@theta[,colnames(simulated_frac)],simulated_frac)) %>% as.data.frame() %>% rownames_to_column('cell_type')
updated.corr=diag(cor(bp.res@posterior.theta_f@theta[,colnames(simulated_frac)],simulated_frac)) %>% as.data.frame() %>% rownames_to_column('cell_type')

df=rbind(initial.corr,updated.corr)  %>% rename('correlation'='.') %>% mutate(method=c(rep('with initial reference',9),rep('with updated reference',9)))

ggplot(df,aes(x = cell_type, y = correlation, group = method)) +
  geom_line(aes(color=method))+
  geom_point(aes(color=method))+
  theme_bw()+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))

```
As can be found here, using the updated reference does not improve the performance significantly, and even leads to performance drop in some cases. Therefore in real practice, to accelerate the deconvolution process as well as to maintain good results, we recommend the users to set `update=F` (which is the default setting in `InstaPrism()`).
\

## Additional notes on InstaPrism parameters
So far we have demonstrated the primary usage of **InstaPrism**, mostly with the default parameters. In real practice, some of the default parameters can be tailored to meet the users' specific requirements.

* **n.iter**: number of iterations in InstaPrism algorithm.  
`n.iter` is an important parameter for InstaPrism algorithm. InstaPrism updates the fraction estimates over iterations and `n.iter` determines how many iterations are performed. Under default setting, this value is set to *max(20, number of cell.states $\times$ 2)*. However, users can set it to larger values to improve convergence. Specifically, we included two additional parameters that help to verify the convergence status: *verbose* and *convergence.plot*.

* **verbose**: a logical variable to determine whether to display convergence status of the model.   
By setting `verbose = T`, `InstaPrsim()` will print the convergence status summary for both cell.states and cell.types.  
Specifically, the absolute difference in fraction estimates between the last two iterations is utilized as an indicator of convergence (abs_diff), with smaller values indicating convergence and higher confidence of the deconvolution results. 
Using `example3.RData` as a demonstration, by setting `verbose = T`, we get a summarized table showing the minimum, median and maximum abs_diff for all cell.states/types across the samples. Usually we consider abs_diff < 0.01 as convergence. 

```{r}
load('extdata/tutorial_example3/example3.RData')
InstaPrism.res = InstaPrism(input_type = 'raw',sc_Expr = scExpr_train,bulk_Expr = simulated_bulk,
                            cell.type.labels = cell.type.labels, cell.state.labels = cell.state.labels, 
                            n.iter = 100,n.core = 16,verbose = T)
```

* **convergence.plot**: a logical variable determining whether to visualize convergence status for cell types.   
Again, using `example3.RData` as a demonstration, first let's consider the scenario when`n.iter` is set too low. According to the convergence plot, fraction estimates for malignant cells did not converge. This means that we should set higher value of `n.iter`.
```{r,fig.align='center',fig.width=10,fig.height=4.5,results='hide'}
InstaPrism.res = InstaPrism(input_type = 'raw',sc_Expr = scExpr_train,bulk_Expr = simulated_bulk,
                            cell.type.labels = cell.type.labels, cell.state.labels = cell.state.labels, 
                            n.iter = 20,n.core = 16,convergence.plot=T,max_n_per_plot = 100)
```

Now, let's try using a higher value of `n.iter`. 
This time we get fraction estimation for all the cell types converge (with all the abs_diff less than 0.01). 
```{r,fig.align='center',fig.width=10,fig.height=4.5,results='hide'}
InstaPrism.res = InstaPrism(input_type = 'raw',sc_Expr = scExpr_train,bulk_Expr = simulated_bulk,
                            cell.type.labels = cell.type.labels, cell.state.labels = cell.state.labels, 
                            n.iter = 100,n.core = 16,convergence.plot=T,max_n_per_plot = 100)
```


* **return.Z.cs**: a logical variable determining whether to return cell.state specific gene expression.   
By default, this parameter is set to FALSE to save memory in the calculation. We recommend that readers keep this parameter at its default setting, particularly when dealing with a large number of distinct cell states.

\
\
\

